### ISSUE TITLE
Handle filesystem destroy / rename

### ISSUE THREAD

**Issue by ahrens** [zrepl/zrepl#70]
We should figure out what should happen when a filesystem is destroyed, or renamed, on the source system.  We could also destroy/rename it on the target.

**Comment by jbliesener** [comment]
As of 0.1.0-rc3, in a pull-job, **the whole pruning phase fails** (for the remaining filesystems as well!), because zrepl can not get the replication cursor bookmark for the deleted filesystem. Left me with 25000 unpruned snapshots, until I found out.

In my specific case, I'm replicating my Proxmox VMs hard disks every 10 minutes. Each VM hard disk is a zvol and therefore a separate dataset. Removing a VM and it's hard disk in Proxmox eliminates the dataset, including snapshots and bookmarks. Consequently, zrepl runs into trouble, because it can't find the cursor bookmark anymore - and aborts the whole pruning process with a PermErr.

In my case, I would expect the existing snapshots to remain on the receiver side and to be pruned according to the pruning sieves. That would leave me with some backups until the period defined in the last pruning sieve.

**Comment by problame** [comment]
The vague idea about this is to move away from sender-side filesystem names and assign GUIDs to filesystems on the sender instead, i.e. property `zrepl:fs_guid`.

We would then use that GUID for replication, as we do with the snapshot  GUID to identify snapshots.

@ahrens I just found out that `filesystem` also has a `guid` property on ZoL (`zfs get guid zroot`). I contributed docs for `guid` on snapshots and bookmarks a while ago ( https://github.com/zfsonlinux/zfs/pull/6102 ). Should I do that for filesystems and zvols, too?

**Comment by wyattanderson** [comment]
Using GUIDs instead of the dataset name would be fantastic for my use case. I'm using Proxmox with ZFS, and the datasets are created using the Proxmox VM number. I _wish_ Proxmox would use GUIDs instead of a sequential number for many reasons, not the least of which is the fact that Proxmox reuses VM numbers.

If you delete a VM with e.g. ID 101, the next VM you create will now have ID 101, and zrepl will balk at the (very real) discrepancy. Cleanup of this situation is also a pain in the ass, as an operator needs to go in to the target machine, remove the holds on the snapshots of the now-deleted VM volumes, and manually delete the volumes. With enough churn in VMs, especially in a lab environment, this becomes disruptive.

**Comment by problame** [comment]
> I wish Proxmox would use GUIDs instead of a sequential number for many reasons, not the least of which is the fact that Proxmox reuses VM numbers.

Wow, I feel your pain.

> Cleanup of this situation is also a pain in the ass, as an operator needs to go in to the target machine, remove the holds on the snapshots of the now-deleted VM volumes, 

You can have a look at the `zrepl zfs-abstraction` subcommand for that. Please be careful, it's a pretty sharp tool.
If you find it useful, please tell me where we can promote it in the docs.

> and manually delete the volumes. With enough churn in VMs, especially in a lab environment, this becomes disruptive.

Let's think about how zrepl can deal with situations like this in general:

* **send-side destroy**: the send-side fs was destroyed, the receive-side fs still exists
  * current zrepl behavior is to leave the receive-side fs alone
* **send-side destroy-then-recreate**: send-side destroyed the fs, then re-created a new one with the same name
  * current zrepl behavior is to call out a replication conflict and give up on that fs
  * zrepl doesn't understand that it's a different filesystem because ATM we use *names*, not fs *guid*s for identity
* **send-side rename to unreplicated name**: send-side fs was renamed to a new fs name that has not been replicated (i.e. doesn't exist on receiver
  * current zrepl behavior is to do an initial replication of the filesystem
* **send-side rename to a previously replicated name**: the new fs name has already been replicated previously, but the previous send-side fs with that name has been either deleted or renamed
  * current zrepl behavior is like **send-side destroy-then-recreate** (indistinguishable)

I'm very willing to assign GUIDs to filesystems on the sending side, replicate them to the receiver, and use them in the future to match up send-side and recv-side filesystems.
What's left to do / design (**calling for input / help here** (cc @InsanePrawn )):

* [ ] Is the list of situations above exhaustive?
* [ ] Can we detect & distinguish all of the above situations with that information?
  Or do we need to track more information per FS & side?
  The code would get the following data as input:
  ```
  type Filesystem interface {
    LocalNameWithoutZreplRootfsPrefix() string // local zfs name property. If it's a receiving side job that has a `root_fs` config variable, that prefix is trimmed.
    LocalGuid() string // local zfs guid property (does not survive send  / recv)
    ZreplGuid() string // zfs zrepl:guid property (will be established on the receiver as part of a zrepl initial replication)
  }

  var senderFilesystems, receiverFilesystems []Filesystem
  ```
  Please quote each of the situations outlined above individually and explain why (or why not) the data outlined above is sufficient.
* [ ] Migration path for existing replication setups
  * Proposal: if the receiving side doesn't have a GUID, fall back to the name as we do now.

**Comment by InsanePrawn** [comment]
Ah, the GUID discussion. :) Guess it would've been helpful if ZFS' own GUIDs were usable for this.
Here's an unstructured braindump from my side:
- enhancing zrepl's existing behaviour through an additional GUID for conflict detection and resolution sounds good to me!
- the list lacks the combination of destroy `poolio/a1` and rename `poolio/b1 -> poolio/a1`, although it's similar to destroy-then-recreate.
What about "zrepl GUID deleted" and even worse "zrepl GUID overwritten by user". What happens when we have a GUID conflict?
And a superset of that list for the recv side.
- Recv side: Falling back to names and storing the GUID after a successful replication sounds like the way to go for me.
This enables admins to circumvent the initial replication, e.g. to restore from some other, potentially faster source.
(Otherwise?) a 'fill missing GUIDs' command will be necessary.
- We'll need to handle garbage collection. E.g. create, use and destroy 'testvm123' a couple times, your backup will now contain a couple trash datasets, unless your chosen conflict resolution path is 'destroy to replace'.
- Which brings me to my next point: Conflict resolution strategies will probably have to be configurable.
  - whether to destroy, rename or respect (as in skip replication of this DS like before) conflicting datasets on the receiving side
  - where to rename to (recycle bin for datasets including pruning?): -> naming scheme?
  - what to do with possible sender-side deletions (detected symptom: receiver has dataset with zrepl GUID, dataset doesn't exist on sender, neither by name nor GUID)
  - fallback behaviour: Some users might prefer to throw problematic datasets away on the recv side and do a fresh repl when the automatic conflict resolution fails (think fun things like encryption hierarchies) or is undecided what to do.

This feature will solve phasing out sender-side deleted datasets. (Currently those datasets fall out of snapshot(!) pruning, plus we currently have no mechanism to clean up "empty" (as in no more snaps to be kept) recv side datasets)

At a glance the `Filesystem` struct LGTM for detecting the scenarios you described.

We should probably plan this to be useful enough for clone handling as well, even if we don't implement the functionality yet. #71
I suspect the proposed struct plus a future optional parent-Filesystem and -Version (snapshot) field will do?

**Comment by problame** [comment]
> * the list lacks the combination of destroy poolio/a1 and rename poolio/b1 -> poolio/a1, although it's similar to destroy-then-recreate.

Added it to the list.

> *
>   * whether to destroy, rename or respect (as in skip replication of this DS like before) conflicting datasets on the receiving side
>   * where to rename to (recycle bin for datasets including pruning?): -> naming scheme?

The go-to smart conflict resolution would probably look like this:
If recv-side fs `a` conflicts (different zrepl guids`), then
* rename recv-side `a` to `a.zrepl.conflict.${a.zrepl_guid}`
* do initial replication of send-side `a` to recv-side
* move recv-side child filesystems `a.zrepl.conflict.${a.zrepl_guid}/*` to recv-side `a/`
  * What a pity there isn't a `zfs rename-swap` command that can do this step in one txg.
    @InsanePrawn  we already talked about this a lot when discussing initial encrypted recvs into zrepl placeholder filesystems. 

> 
This feature will solve phasing out sender-side deleted datasets. (Currently those datasets fall out of snapshot(!) pruning, plus we currently have no mechanism to clean up "empty" (as in no more snaps to be kept) recv side datasets)

You mean recv-side datasets?
That's probably orthogonal.
Even with zrepl's current design, we could make a pruner that destroys a filesystem that doesn't `keep` any snapshots.
That behavior would only be enabled on a receiving-side job, not a sender.
Basically like: if the list of snapshots of a recv-side filesystem, we won't ever be able to recv into it anymore, might as well destroy it.

> We should probably plan this to be useful enough for clone handling as well, even if we don't implement the functionality yet. #71
> I suspect the proposed struct plus a future optional parent-Filesystem and -Version (snapshot) field will do?

I still think a flat, GUID-based naming scheme for the recv-side filesystems would save us a lot of trouble here...

**Comment by problame** [comment]
Like this:
![image](https://user-images.githubusercontent.com/956573/93628144-60cc0880-f9e6-11ea-9ddb-953008129c7c.png)

**Comment by problame** [comment]
@wyattanderson As an administrator with your Proxmox setup, how would you feel of this kind of `sink`-side filesystem hierarchy layout?

**Comment by wyattanderson** [comment]
Looks good to me, but I don't have enough expertise with ZFS and the usecases of `zrepl` to comment much more beyond that.

**Comment by problame** [comment]
Well, for instance, his do you feel about the different ideas for what we should do with datasets on the receiver that were deleted on the sender? What would you expect to happen?
